tensorflow/core/kernels/data/concatenate_dataset_op_test.cc

```diff
@@ -0,0 +1,596 @@
+                                               &tensor_slice_dataset_tensors));
+  gtl::InlinedVector<TensorValue, 4> inputs;
+  for (auto &tensor : tensor_slice_dataset_tensors) {
+    inputs.emplace_back(&tensor);
+  }
+  std::vector<PartialTensorShape> output_shapes =
+      TestCases[0].expected_output_shapes;
+  DataTypeVector output_dtypes = TestCases[0].expected_output_dtypes;
+  std::unique_ptr<OpKernel> dataset_kernel;
+  TF_ASSERT_OK(CreateConcatenateDatasetKernel(output_dtypes, output_shapes,
+                                              &dataset_kernel));
+  std::unique_ptr<OpKernelContext> dataset_kernel_ctx;
+  TF_ASSERT_OK(CreateConcatenateDatasetContext(dataset_kernel.get(), &inputs,
+                                               &dataset_kernel_ctx));
+  DatasetBase *dataset;
+  TF_ASSERT_OK(
+      CreateDataset(dataset_kernel.get(), dataset_kernel_ctx.get(), &dataset));
+  core::ScopedUnref scoped_unref(dataset);
+
+  EXPECT_EQ(dataset->name(), kOpName);
+}
+
+struct DatasetOutputDtypesTest : ConcatenateDatasetOpTest,
+                                 ::testing::WithParamInterface<TestParam> {};
+
+TEST_P(DatasetOutputDtypesTest, DatasetOutputDtypes) {
+  int thread_num = 2, cpu_num = 2;
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
+
+  std::vector<std::vector<Tensor>> tensor_vectors = GetParam().input_tensors;
+  std::vector<Tensor> expected_outputs = GetParam().expected_outputs;
+  std::vector<PartialTensorShape> expected_output_shapes =
+      GetParam().expected_output_shapes;
+  DataTypeVector expected_output_dtypes = GetParam().expected_output_dtypes;
+  std::vector<Tensor> tensor_slice_dataset_tensors;
+  TF_ASSERT_OK(CreateTensorSliceDatasetTensors(tensor_vectors,
+                                               &tensor_slice_dataset_tensors));
+  gtl::InlinedVector<TensorValue, 4> inputs;
+  for (auto &tensor : tensor_slice_dataset_tensors) {
+    inputs.emplace_back(&tensor);
+  }
+
+  std::unique_ptr<OpKernel> dataset_kernel;
+  TF_ASSERT_OK(CreateConcatenateDatasetKernel(
+      expected_output_dtypes, expected_output_shapes, &dataset_kernel));
+  std::unique_ptr<OpKernelContext> dataset_kernel_ctx;
+  TF_ASSERT_OK(CreateConcatenateDatasetContext(dataset_kernel.get(), &inputs,
+                                               &dataset_kernel_ctx));
+  DatasetBase *dataset;
+  TF_ASSERT_OK(
+      CreateDataset(dataset_kernel.get(), dataset_kernel_ctx.get(), &dataset));
+  core::ScopedUnref scoped_unref(dataset);
+
+  EXPECT_OK(VerifyTypesMatch(dataset->output_dtypes(), expected_output_dtypes));
+}
+
+INSTANTIATE_TEST_CASE_P(ConcatenateDatasetOpTest, DatasetOutputDtypesTest,
+                        ::testing::ValuesIn(TestCases));
+
+struct DatasetOutputShapesTest : ConcatenateDatasetOpTest,
+                                 ::testing::WithParamInterface<TestParam> {};
+
+TEST_P(DatasetOutputShapesTest, DatasetOutputShapes) {
```

Comment:
Is it possible to reduce the code repetition here? Many of these tests have the same general structure.

Reply:
@rachellim There is a solution as below. If it works, I will submit a commit to change the other tests.

cc @jsimsa : I will also revise the other two pending PRs (#25599 and #25496) after the solution is finalized.


```C++
struct ConcatenateDatasetOpTestHelper
    : ConcatenateDatasetOpTest,
      ::testing::WithParamInterface<TestParam> {
  Status CreateDatasetFromTestParam() {
    std::vector<Tensor> tensor_slice_dataset_tensors;
    TF_RETURN_IF_ERROR(CreateTensorSliceDatasetTensors(
        tensor_vectors, &tensor_slice_dataset_tensors));
    gtl::InlinedVector<TensorValue, 4> inputs;
    for (auto &tensor : tensor_slice_dataset_tensors) {
      inputs.emplace_back(&tensor);
    }

    TF_RETURN_IF_ERROR(CreateConcatenateDatasetKernel(
        expected_output_dtypes, expected_output_shapes, &dataset_kernel));
    TF_RETURN_IF_ERROR(CreateConcatenateDatasetContext(
        dataset_kernel.get(), &inputs, &dataset_kernel_ctx));
    TF_RETURN_IF_ERROR(CreateDataset(dataset_kernel.get(),
                                     dataset_kernel_ctx.get(), &dataset));
  }

  Status CreateIteratorFromTestParam() {
    TF_RETURN_IF_ERROR(
        CreateIteratorContext(dataset_kernel_ctx.get(), &iterator_ctx));
    TF_RETURN_IF_ERROR(
        dataset->MakeIterator(iterator_ctx.get(), "Iterator", &iterator));
  }

  const std::vector<std::vector<Tensor>> &tensor_vectors =
      GetParam().input_tensors;
  const std::vector<Tensor> &expected_outputs = GetParam().expected_outputs;
  std::vector<Tensor>::const_iterator expected_outputs_it =
      expected_outputs.begin();
  const std::vector<PartialTensorShape> &expected_output_shapes =
      GetParam().expected_output_shapes;
  const DataTypeVector &expected_output_dtypes =
      GetParam().expected_output_dtypes;
  std::unique_ptr<OpKernel> dataset_kernel;
  std::unique_ptr<OpKernelContext> dataset_kernel_ctx;
  DatasetBase *dataset;
  std::unique_ptr<IteratorContext> iterator_ctx;
  std::unique_ptr<IteratorBase> iterator;
};

struct DatasetGetNextTest : public ConcatenateDatasetOpTestHelper {};

TEST_P(DatasetGetNextTest, GetNext) {
  int thread_num = 2, cpu_num = 2;
  TF_ASSERT_OK(InitThreadPool(thread_num));
  TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
  TF_ASSERT_OK(CreateDatasetFromTestParam());
  core::ScopedUnref scoped_unref(dataset);
  TF_ASSERT_OK(CreateIteratorFromTestParam());

  bool end_of_sequence = false;
  std::vector<Tensor> out_tensors;
  while (!end_of_sequence) {
    TF_EXPECT_OK(
        iterator->GetNext(iterator_ctx.get(), &out_tensors, &end_of_sequence));
    if (!end_of_sequence) {
      for (auto &tensor : out_tensors) {
        EXPECT_NE(expected_outputs_it, expected_outputs.end());
        TF_EXPECT_OK(ExpectEqual(tensor, *expected_outputs_it));
        expected_outputs_it++;
      }
    }
  }
  EXPECT_EQ(expected_outputs_it, expected_outputs.end());
}

INSTANTIATE_TEST_CASE_P(ConcatenateDatasetOpTest, DatasetGetNextTest,
                        ::testing::ValuesIn(TestCases));
```